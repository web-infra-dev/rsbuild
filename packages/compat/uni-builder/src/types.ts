import type {
  NodeEnv,
  MetaOptions,
  ScriptInject,
  RsbuildTarget,
  ChainedConfig,
  ChainedConfigWithUtils,
  InlineChunkTest,
} from '@rsbuild/shared';
import type { RsbuildConfig as RsbuildRspackConfig } from '@rsbuild/core';
import type { RsbuildConfig as RsbuildWebpackConfig } from '@rsbuild/webpack';
import type { PluginAssetsRetryOptions } from '@rsbuild/plugin-assets-retry';
import type { LazyCompilationOptions } from './webpack/plugins/lazyCompilation';
import type { PluginRemOptions } from '@rsbuild/plugin-rem';

export type CreateWebpackBuilderOptions = {
  bundlerType: 'webpack';
  config: UniBuilderWebpackConfig;
  frameworkConfigPath?: string;
  /** The root path of current project. */
  cwd: string;
};

export type CreateRspackBuilderOptions = {
  bundlerType: 'rspack';
  config: UniBuilderRspackConfig;
  frameworkConfigPath?: string;
  /** The root path of current project. */
  cwd: string;
};

export type CreateUniBuilderOptions =
  | CreateWebpackBuilderOptions
  | CreateRspackBuilderOptions;

export type RsbuildConfig<B = 'rspack'> = B extends 'rspack'
  ? RsbuildRspackConfig
  : RsbuildWebpackConfig;

export type GlobalVars = Record<string, any>;

export type ChainedGlobalVars = ChainedConfigWithUtils<
  GlobalVars,
  {
    env: NodeEnv;
    target: RsbuildTarget;
  }
>;

export type ModuleScopes = Array<string | RegExp>;

export type MainFields = (string | string[])[];

export type DevServerHttpsOptions = boolean | { key: string; cert: string };

export type UniBuilderExtraConfig = {
  dev?: {
    /**
     * Used to set the host of Dev Server.
     */
    host?: string;
    /**
     * After configuring this option, you can enable HTTPS Dev Server, and disabling the HTTP Dev Server.
     */
    https?: DevServerHttpsOptions;
    /**
     * Specify a port number for Dev Server to listen.
     */
    port?: number;
  };
  source?: {
    /**
     * Define global variables. It can replace expressions like `process.env.FOO` in your code after compile.
     */
    globalVars?: ChainedGlobalVars;
    /**
     * Restrict importing paths. After configuring this option, all source files can only import code from
     * the specific paths, and import code from other paths is not allowed.
     */
    moduleScopes?: ChainedConfig<ModuleScopes>;
    /**
     * This configuration will determine which field of `package.json` you use to import the `npm` module.
     * Same as the [resolve.mainFields](https://webpack.js.org/configuration/resolve/#resolvemainfields) config of webpack.
     */
    resolveMainFields?: MainFields | Partial<Record<RsbuildTarget, MainFields>>;
    /**
     * Add a prefix to [resolve.extensions](https://webpack.js.org/configuration/resolve/#resolveextensions).
     */
    resolveExtensionPrefix?: string | Partial<Record<RsbuildTarget, string>>;
  };
  output?: {
    /**
     * @deprecated use `output.cssModules.localIdentName` instead
     */
    cssModuleLocalIdentName?: string;
    /**
     * Whether to generate a manifest file that contains information of all assets.
     */
    enableAssetManifest?: boolean;
    /**
     * Configure the retry of assets.
     */
    assetsRetry?: PluginAssetsRetryOptions;
    /**
     * Controls whether to the inline the runtime chunk to HTML.
     */
    disableInlineRuntimeChunk?: boolean;
    /**
     * Convert px to rem in CSS.
     */
    convertToRem?: boolean | PluginRemOptions;
    /**
     * Whether to treat all .css files in the source directory as CSS Modules.
     */
    disableCssModuleExtension?: boolean;
    /**
     * If this option is enabled, all unrecognized files will be emitted to the dist directory.
     * Otherwise, an exception will be thrown.
     */
    enableAssetFallback?: boolean;
    /**
     * @deprecated use `output.inlineScripts` instead
     */
    enableInlineScripts?: boolean | InlineChunkTest;
    /**
     * @deprecated use `output.inlineStyles` instead
     */
    enableInlineStyles?: boolean | InlineChunkTest;
  };
  html?: {
    /**
     * @deprecated use `html.meta` instead
     */
    metaByEntries?: Record<string, MetaOptions>;
    /**
     * @deprecated use `html.title` instead
     */
    titleByEntries?: Record<string, string>;
    /**
     * @deprecated use `html.favicon` instead
     */
    faviconByEntries?: Record<string, string | undefined>;
    /**
     * @deprecated use `html.inject` instead
     */
    injectByEntries?: Record<string, ScriptInject>;
    /**
     * @deprecated use `html.template` instead
     */
    templateByEntries?: Partial<Record<string, string>>;
    /**
     * @deprecated use `html.templateParameters` instead
     */
    templateParametersByEntries?: Record<string, Record<string, unknown>>;
  };
};

export type SriOptions = {
  hashFuncNames?: [string, ...string[]];
  enabled?: 'auto' | true | false;
  hashLoading?: 'eager' | 'lazy';
};

export type UniBuilderWebpackConfig = RsbuildWebpackConfig &
  UniBuilderExtraConfig & {
    security?: {
      /**
       * Adding an integrity attribute (`integrity`) to sub-resources introduced by HTML allows the browser to
       * verify the integrity of the introduced resource, thus preventing tampering with the downloaded resource.
       */
      sri?: SriOptions | boolean;
    };
    experiments?: {
      lazyCompilation?: LazyCompilationOptions;
    };
  };

export type UniBuilderRspackConfig = RsbuildRspackConfig &
  UniBuilderExtraConfig;

export type BuilderConfig<B = 'rspack'> = B extends 'rspack'
  ? UniBuilderRspackConfig
  : UniBuilderWebpackConfig;
