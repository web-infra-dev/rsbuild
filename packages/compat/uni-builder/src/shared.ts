import { deepmerge } from '@rsbuild/shared/deepmerge';
import type {
  RsbuildPlugin,
  RsbuildConfig as RsbuildRspackConfig,
} from '@rsbuild/core';
import type { RsbuildConfig as RsbuildWebpackConfig } from '@rsbuild/webpack';
import type { UniBuilderRspackConfig, UniBuilderWebpackConfig } from './types';

export function parseCommonConfig<B = 'rspack' | 'webpack'>(
  uniBuilderConfig: B extends 'rspack'
    ? UniBuilderRspackConfig
    : UniBuilderWebpackConfig,
): {
  rsbuildConfig: B extends 'rspack'
    ? RsbuildRspackConfig
    : RsbuildWebpackConfig;
  rsbuildPlugins: RsbuildPlugin[];
} {
  const rsbuildConfig = deepmerge({}, uniBuilderConfig);
  const { output } = rsbuildConfig;

  if (output.cssModuleLocalIdentName) {
    output.cssModules ||= {};
    output.cssModules.localIdentName = output.cssModuleLocalIdentName;
    delete output.cssModuleLocalIdentName;
  }

  return {
    rsbuildConfig: rsbuildConfig,
    rsbuildPlugins: [],
  };
}
