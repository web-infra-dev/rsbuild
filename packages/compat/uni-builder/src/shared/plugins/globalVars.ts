import {
  mergeChainedOptions,
  type Define,
  type DefaultRsbuildPlugin,
} from '@rsbuild/shared';
import type { ChainedGlobalVars } from '../../types';

export const pluginGlobalVars = (
  options?: ChainedGlobalVars,
): DefaultRsbuildPlugin => ({
  name: 'plugin-global-vars',

  setup(api) {
    api.modifyBundlerChain((chain, { env, target, bundler }) => {
      if (!options) {
        return;
      }

      const globalVars = mergeChainedOptions({
        defaults: {},
        options,
        utils: { env, target },
      });

      const serializedVars: Define = {};

      Object.entries(globalVars).forEach(([key, value]) => {
        serializedVars[key] = JSON.stringify(value) ?? 'undefined';
      });

      chain.plugin('globalVars').use(bundler.DefinePlugin, [serializedVars]);
    });
  },
});
