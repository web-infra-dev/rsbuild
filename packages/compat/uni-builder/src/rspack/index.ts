import { createRsbuild } from '@rsbuild/core';
import type { RsbuildInstance, RsbuildPlugin } from '@rsbuild/core';
import type {
  RsbuildConfig,
  RspackProvider,
} from '@rsbuild/core/rspack-provider';
import type { UniBuilderRspackConfig } from '../types';
import type { CreateRspackBuilderOptions } from '../types';
import { parseCommonConfig } from '../shared/parseCommonConfig';

export async function parseConfig(
  uniBuilderConfig: UniBuilderRspackConfig,
  frameworkConfigPath?: string,
): Promise<{
  rsbuildConfig: RsbuildConfig;
  rsbuildPlugins: RsbuildPlugin[];
}> {
  const { rsbuildConfig, rsbuildPlugins } = parseCommonConfig<'rspack'>(
    uniBuilderConfig,
    frameworkConfigPath,
  );

  if (uniBuilderConfig.output?.enableAssetManifest) {
    const { pluginManifest } = await import('./plugins/manifest');
    rsbuildPlugins.push(pluginManifest());
  }

  return {
    rsbuildConfig,
    rsbuildPlugins,
  };
}

export async function createRspackBuilder(
  options: CreateRspackBuilderOptions,
): Promise<RsbuildInstance<RspackProvider>> {
  const { rsbuildConfig, rsbuildPlugins } = await parseConfig(
    options.config,
    options.frameworkConfigPath,
  );
  const rsbuild = await createRsbuild({
    rsbuildConfig,
  });

  rsbuild.addPlugins(rsbuildPlugins);

  return rsbuild;
}
