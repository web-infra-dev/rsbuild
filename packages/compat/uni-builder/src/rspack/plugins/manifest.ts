import type { RsbuildPlugin, RsbuildPluginAPI } from '@rsbuild/core';
import { generateManifest } from '../../shared/manifest';

export const pluginManifest = (): RsbuildPlugin<RsbuildPluginAPI> => ({
  name: 'plugin-manifest',

  setup(api) {
    api.modifyBundlerChain(async (chain, { CHAIN_ID }) => {
      const { WebpackManifestPlugin } = await import('rspack-manifest-plugin');
      const publicPath = chain.output.get('publicPath');

      chain.plugin(CHAIN_ID.PLUGIN.MANIFEST).use(WebpackManifestPlugin, [
        {
          fileName: 'asset-manifest.json',
          publicPath,
          generate: generateManifest,
        },
      ]);
    });
  },
});
