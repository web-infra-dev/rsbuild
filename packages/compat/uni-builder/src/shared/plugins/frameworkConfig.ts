import { fse, type DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginFrameworkConfig = (
  configPath: string,
): DefaultRsbuildPlugin => ({
  name: 'plugin-framework-config',

  setup(api) {
    api.modifyBundlerChain((chain) => {
      if (!fse.existsSync(configPath)) {
        return;
      }

      const cache = chain.get('cache');

      cache.buildDependencies = {
        ...cache.buildDependencies,
        frameworkConfig: [configPath],
      };

      chain.cache(cache);
    });
  },
});
