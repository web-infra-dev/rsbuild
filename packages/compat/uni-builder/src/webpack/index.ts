import {
  createRsbuild,
  type RsbuildConfig,
  type RsbuildPlugin,
  type RsbuildInstance,
} from '@rsbuild/core';
import type { RsbuildTarget } from '@rsbuild/shared';
import type { UniBuilderWebpackConfig } from '../types';
import type { CreateWebpackBuilderOptions } from '../types';
import { parseCommonConfig } from '../shared/parseCommonConfig';
import { pluginModuleScopes } from './plugins/moduleScopes';
import { pluginStyledComponents } from './plugins/styledComponents';
import { pluginBabel } from './plugins/babel';
import { pluginReact } from './plugins/react';
import { withDefaultConfig } from './defaults';

export async function parseConfig(
  uniBuilderConfig: UniBuilderWebpackConfig,
  cwd: string,
  frameworkConfigPath?: string,
  target?: RsbuildTarget | RsbuildTarget[],
): Promise<{
  rsbuildConfig: RsbuildConfig;
  rsbuildPlugins: RsbuildPlugin[];
}> {
  const { rsbuildConfig, rsbuildPlugins } = await parseCommonConfig<'webpack'>(
    uniBuilderConfig,
    cwd,
    frameworkConfigPath,
    target,
  );

  rsbuildPlugins.push(
    pluginBabel({
      babelLoaderOptions: uniBuilderConfig.tools?.babel,
    }),
  );
  rsbuildPlugins.push(pluginReact());

  if (uniBuilderConfig.tools?.tsLoader) {
    const { pluginTsLoader } = await import('./plugins/tsLoader');
    rsbuildPlugins.push(
      pluginTsLoader(
        uniBuilderConfig.tools.tsLoader,
        uniBuilderConfig.tools.babel,
      ),
    );
  }

  if (uniBuilderConfig.output?.enableAssetManifest) {
    const { pluginManifest } = await import('./plugins/manifest');
    rsbuildPlugins.push(pluginManifest());
  }

  if (uniBuilderConfig.security?.sri) {
    const { pluginSRI } = await import('./plugins/sri');
    rsbuildPlugins.push(pluginSRI(uniBuilderConfig.security?.sri));
  }

  if (uniBuilderConfig.experiments?.lazyCompilation) {
    const { pluginLazyCompilation } = await import('./plugins/lazyCompilation');
    rsbuildPlugins.push(
      pluginLazyCompilation(uniBuilderConfig.experiments?.lazyCompilation),
    );
  }

  rsbuildPlugins.push(
    pluginStyledComponents(uniBuilderConfig.tools?.styledComponents),
  );

  return {
    rsbuildConfig,
    rsbuildPlugins,
  };
}

export async function createWebpackBuilder(
  options: CreateWebpackBuilderOptions,
): Promise<RsbuildInstance> {
  const { rsbuildConfig, rsbuildPlugins } = await parseConfig(
    withDefaultConfig(options.config),
    options.cwd,
    options.frameworkConfigPath,
    options.target,
  );

  const { webpackProvider } = await import('@rsbuild/webpack');
  const rsbuild = await createRsbuild({
    rsbuildConfig,
    provider: webpackProvider,
    target: options.target || 'web',
  });

  rsbuild.addPlugins([
    ...rsbuildPlugins,
    pluginModuleScopes(options.config.source?.moduleScopes),
  ]);

  return rsbuild;
}
