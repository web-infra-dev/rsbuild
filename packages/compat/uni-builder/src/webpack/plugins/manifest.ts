import type { RsbuildPlugin } from '@rsbuild/webpack';
import { generateManifest } from '../../shared/manifest';

export const pluginManifest = (): RsbuildPlugin => ({
  name: 'uni-builder:manifest',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const { WebpackManifestPlugin } = await import('webpack-manifest-plugin');
      const publicPath = chain.output.get('publicPath');

      chain.plugin(CHAIN_ID.PLUGIN.MANIFEST).use(WebpackManifestPlugin, [
        {
          fileName: 'asset-manifest.json',
          publicPath,
          generate: generateManifest,
        },
      ]);
    });
  },
});
