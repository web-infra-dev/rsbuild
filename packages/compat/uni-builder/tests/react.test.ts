import { expect, describe, it } from 'vitest';
import { pluginReact } from '../src/webpack/plugins/react';
import { webpackProvider } from '../../webpack/src';
import { pluginBabel } from '../src/webpack/plugins/babel';
import { createStubRsbuild } from '@rsbuild/test-helper';

describe('plugins/react', () => {
  it('should work with babel-loader', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel(), pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {},
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not apply react refresh when dev.hmr is false', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {
        dev: {
          hmr: false,
        },
      },
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is node', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {
        output: {
          targets: ['node'],
        },
      },
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is web-worker', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {
        output: {
          targets: ['web-worker'],
        },
      },
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });
});
