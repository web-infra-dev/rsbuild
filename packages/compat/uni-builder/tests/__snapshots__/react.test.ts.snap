// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/react > should work with babel-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
        ],
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/uni-builder/compiled/babel-plugin-lodash/index.js",
                  {},
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/react-refresh/babel.js",
                  {
                    "skipEnvCheck": true,
                  },
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/uni-builder/compiled/babel-plugin-lodash/index.js",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    ReactRefreshPlugin {
      "options": {
        "exclude": [
          /node_modules/,
        ],
        "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
        "overlay": false,
      },
    },
  ],
}
`;
